### PR TITLE
Measure, save, and plot progress and traffic statistics

### PR DESCRIPTION
**Pull request by cskiraly**
- Keep track of progress and traffic statistics
- Save these as vectors (if saveProgress)
- plot for each single run (if plotProgress)

### CODE REVIEW

File: DAS/simulator.py [hunk 1]
```diff
@@ -213,6 +239,20 @@ def run(self):
             else:
                 steps += 1
 
+        progress = pd.DataFrame(progressVector)
```
[the remark this answers is not in the document]
Comment (cskiraly) [reply DAS/simulator.py:242]: Good point. I can skip the plotting code from the PR, and we add later?

File: DAS/simulator.py [hunk 2]
```diff
@@ -213,6 +239,20 @@ def run(self):
...
+        if self.config.saveProgress:
+            self.result.addMetric("progress", progress.to_dict(orient='list'))
+        if self.config.plotProgress:
+            progress.plot.line(subplots = [[cnS, cnN, cnV], [cnT0], [cnT1, cnR1, cnD1], [cnT2, cnR2, cnD2]],
+                            title = str(self.shape))
+            if not os.path.exists("results"):
+                os.makedirs("results")
+            if not os.path.exists("results/"+self.execID):
+                os.makedirs("results/"+self.execID)
+            filePath = "results/"+self.execID+"/"+str(self.shape)+".png"
```
[the remark this answers is not in the document]
Comment (cskiraly) [reply DAS/simulator.py:252]: agree. I'm postponing this last patch.

File: DAS/simulator.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (leobago) [reply DAS/simulator.py:242]: Yup, sounds good.

File: DAS/simulator.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (leobago) [reply DAS/simulator.py:252]: Thanks

File: DAS/results.py [hunk 3]
```diff
@@ -26,6 +27,9 @@ def populate(self, shape, missingVector):
             self.blockAvailable = 0
             self.tta = -1
 
+    def addMetric(self, name, metric):
```
[the remark this answers is not in the document]
Comment (cskiraly) [reply DAS/results.py:30]: added

File: DAS/observer.py [hunk 4]
```diff
@@ -44,9 +45,43 @@ def checkStatus(self, validators):
...
                 (a, e) = val.checkStatus()
                 arrived += a
                 expected += e
-        return (arrived, expected)
+                if a == e:
+                    ready += 1
+                    validated += val.vpn
+        return (arrived, expected, ready, validated)
+
+    def getProgress(self, validators):
```
[the remark this answers is not in the document]
Comment (cskiraly) [reply DAS/observer.py:60]: added

File: DAS/observer.py [hunk 5]
```diff
@@ -44,9 +45,43 @@ def checkStatus(self, validators):
...
+            arrived, expected, ready, validated = self.checkStatus(validators)
+            missingSamples = expected - arrived
+            sampleProgress = arrived / expected
+            nodeProgress = ready / (len(validators)-1)
+            validatorCnt = sum([v.vpn for v in validators[1:]])
+            validatorProgress = validated / validatorCnt
+
+            return missingSamples, sampleProgress, nodeProgress, validatorProgress
+
+    def getTrafficStats(self, validators):
```
[the remark this answers is not in the document]
Comment (cskiraly) [reply DAS/observer.py:81]: added

**Comment by leobago** [comment]
LGTM!

**Pull request merged**